### PR TITLE
Cherry pick

### PR DESCRIPTION
**Pull request by KevinRansom**
Fixes: https://github.com/dotnet/fsharp/issues/17161

Cherry pick fix from:  Fixes: https://github.com/dotnet/fsharp/issues/17161

- [x] Add test cases
- [x] Fix it

This is a regression it should go into 8.0.300
/cc @vzarytovskii , @baronfel l did test cases

### CODE REVIEW

**Comment by vzarytovskii** [comment]
Do we need to bump package version, so nuget is uploaded as well?

**Comment by KevinRansom** [comment]
> Do we need to bump package version, so nuget is uploaded as well?

Yes

**Review by T-Gro — APPROVED** [review]

**Comment by vzarytovskii** [comment]
We shouldn't have merged it before tactics approval for the backport

**Pull request merged**
